torch/_inductor/codegen/wrapper.py

```diff
@@ -415,3 +509,145 @@ def call_kernel(self, name: str, kernel: Kernel):
 
     def writeline(self, line):
         self.lines.append(line)
+
+
+class CppWrapperCodeGen(WrapperCodeGen):
+    """
+    The outer wrapper that calls the kernels.
+    """
+
+    call_func_id = count()
+
+    def __init__(self):
+        self._call_func_id = next(CppWrapperCodeGen.call_func_id)
+        super().__init__()
+
+    def write_prefix(self):
+        self.prefix.splice(
+            """
+            async_compile.wait(globals())
+            del async_compile
+            from torch.utils.cpp_extension import load_inline
+            wrapper = (
+            '''
+            #include <dlfcn.h>
+            #include <assert.h>
+            """
+        )
+        with self.wrapper_call.indent():
+            inputs_len = len(V.graph.graph_inputs.keys())
+            output_refs = [x.codegen_reference() for x in V.graph.graph_outputs]
+            if output_refs:
+                if len(output_refs) == 1:
+                    output_types = "at::Tensor"
+                else:
+                    output_return_type = "at::Tensor"
+                    output_return_types = [output_return_type] * len(output_refs)
+                    output_return_types = ", ".join(output_return_types)
+                    output_types = f"std::tuple<{output_return_types}>"
+            else:
+                output_types = "void"
+
+            if inputs_len != 0:
+                inputs_args = ["at::Tensor"] * len(V.graph.graph_inputs.keys())
+                inputs_args = ", ".join(inputs_args)
+                inputs_args = f"std::tuple<{inputs_args}>"
```

Comment:
Doing the args like this will cause an increase in peak memory usage.

The Python wrapper does:
```
def foo(args_as_list):
  a, b = args_as_list
  args_as_list.clear()  # remove references to inputs in caller
  ...
  del a  # first input can now be de-allocated
  ...
  del b  # second input can now be de-allocated
  ....
```

While I believe this version will not de-allocate input buffers until the entire function returns.

I think we can do something similar in this backend, but we may need to use the CPython APIs directly.  If you have a pointer to the PyList object you can call [PyList_SetSlice](https://docs.python.org/2/c-api/list.html#c.PyList_SetSlice) with a NULL arg to do the same thing as `args_as_list.clear()`.

You could look at [guards.cpp](https://github.com/pytorch/pytorch/blob/master/torch/csrc/dynamo/guards.cpp) for an example of working with Tensors + the raw CPython API directly.


cc @Chillee

Reply:
@chunyuan-w Examples like this (https://github.com/pytorch/pytorch/blob/master/torch/csrc/dynamo/guards.cpp#L344) as Jason pointed out should be good to refer to.

Reply:
Also, with that path, we don't have to rely on the slow pt_extension build which uses ninja build...

Reply:
@jansel thanks for the suggestion!
As the `cpp_wrapper` will be turned off by default, may I fix this issue in a separate PR since it will introduce a relatively big change to the current PR? We need to switch from using cpp_extension to CPython.
I've marked this one in the TODOs in the PR description.
If possible, we'd like to firstly have a base implementation of cpp wrapper that is functional.